### PR TITLE
Expose the buffer state as an attribute

### LINKED ISSUE #3013: Try-catching getMappedRange is useful but shouldn't be
If you want to know the map state of a GPUBuffer, you can track the state yourself and update it each time you createBuffer/mapAsync/unmap. This is a fine pattern. However in some cases, if you happen to also need to getMappedRange, you can use the exception thrown from getMappedRange() instead. This isn't great because it (1) encourages authors to depend on exception results, and (2) doesn't allow distinguishing from other errors in getMappedRange() (except by looking at the exception message which is implementation-defined and non-stable).

There are two things we could do about this (either or both):
- Make the state-check in getMappedRange() produce an error in a different way, either return null (#3010) or throw a different exception type.
- Directly expose the state of the buffer as a `readonly attribute` (TBD what the granularity of that state enum is). #3014

### PR DESCRIPTION
**Pull request by kainino0x**
Fixes #3013

***
<a href="https://pr-preview.s3.amazonaws.com/kainino0x/gpuweb/pull/3014.html" title="Last updated on Jun 7, 2022, 9:10 PM UTC (062591b)">Preview</a> | <a href="https://pr-preview.s3.amazonaws.com/gpuweb/gpuweb/3014/f198c8d...kainino0x:062591b.html" title="Last updated on Jun 7, 2022, 9:10 PM UTC (062591b)">Diff</a>

### CODE REVIEW

File: spec/index.bs
Comment (Kangz) [reply spec/index.bs:2468]: Ah, I see this is noted as something to consider in https://github.com/gpuweb/gpuweb/issues/3013#issuecomment-1147989788

Comment (kainino0x) [reply spec/index.bs:2468]: if we do keep them merged like this, then it would be good to come up with a better name. "unmapped" makes it sound like you can use it in a submit, which is wrong.

**Comment by kainino0x** [comment]
Sketched some possibilities on how to expose this
https://docs.google.com/spreadsheets/d/1p5xAB0sh-jAj493RXMeq2Kvbq9LlofEnaGVF86fkcO8/edit?usp=sharing

**Comment by Kangz** [comment]
I thought I remember a discussion that "unmapped" is unclear but couldn't find it again. I made proposal 4 which is proposal 2 but uses the name "unmapped". I think "mappable" would be misleading because some buffers are in the unmapped state but aren't mappable so that enum value would be confusing. IMHO developers will understand that unmapped means "usable on the GPU".

The other thing is that I think we should make the distinction between pending and unmapped because otherwise there isn't a way to know if the buffer can be used in a submit.

**Comment by kainino0x** [comment]
[Editors]
Myles points out there could be a difference in what the application does in response to "not mapped" vs "pending". For example:
- "not mapped" -> try to map it
- "pending" -> wait for mapping

Most likely there isn't any value in trying to hide this detail from the user, so we should probably expose it. That said, I'm in the middle of trying to work through the detangling of content-private vs content-shared vs device-timeline state and that might provide an answer to this issue.

**Comment by kainino0x** [comment]
Another thought: the "locked" state probably isn't necessary. In particular, we wouldn't be guaranteeing that if a buffer is not "locked", the map will actually succeed. It can still fail due to mapAsync validation.

With multithreading, we also don't really need to expose a "locked" state when mapped on another thread - that'll be covered by the existing validation check in mapAsync that checks the buffer isn't mapped already.

So revising this proposal to have three states: "unmapped", "pending", and "mapped".

**Comment by kainino0x** [comment]
Note with this proposal, "destroyed" can be a purely service side state as I have been hoping. In fact, the entire "buffer state" can move to the service side, which I've now gone ahead and done.

Note in the multithreaded future, "unmapped" and "pending" mean just on this thread/this GPUBuffer object. It might also be in any state on another thread.

**Comment by Kangz** [comment]
> So revising this proposal to have three states: "unmapped", "pending", and "mapped".

Revised proposal and skimming the diff, things look good, thanks!

**Review by toji — CHANGES_REQUESTED** [review]
Looks good, but don't we need to transition the `mapState` back to `"unmapped"` in the `.unmap()` algorithm? That's missing here.

**Comment by kainino0x** [comment]
> Looks good, but don't we need to transition the `mapState` back to `"unmapped"` in the `.unmap()` algorithm? That's missing here.

Oops, forgot this. That algorithm is kind of nonsensical right now so we also need to fix that soon. Maybe with #3264.

**Review by toji — APPROVED** [review]

File: spec/index.bs
Comment (kainino0x) [inline spec/index.bs:3014]: Hm, I thought this was a non-change from the existing spec but it's slightly different because this rejects immediately instead of later.

**Comment by kdashg** [comment]
<details><summary><a href="https://github.com/gpuweb/gpuweb/wiki/gpu-web-meeting-2022-08-0203-apac-timed#expose-the-buffer-state-as-an-attribute-3014">WGSL meeting minutes 2022-08-02</a></summary>

* KN: This has gone through a bunch of revisions. I’ve been working through state on client vs server for buffer mapping. What I came to is that client always needs to know whether it’s pending, can’t throw that away, so needs at least three states on client side. Observation from editors meeting from MM, is that it could be useful to know whether it’s mapped vs mappable. Also, it’s probably not worth hiding this from authors, and we should just let them check.
* KN: So unmapped/pending/mapped should work well, shouldn’t be onerous, and all three should be useful for devs, so that’s my proposal.
* KN: For context, in #3013, our resolution was that we should probably just expose this, so this proposal is that we just expose this.
* MM: Discussing this internally. We feel that it this is tracked on client, it’s useful to expose the enum, rather than a more arcane way to get this info.
* MM: Another option here (not advocating particularly) is that we don’t have any state on the client.
* KN: Sort of jives with some stuff I’m considering but I don’t think it’s possible.
* KG: I think we determined in Firefox it was possible but onerous. Cases where if you map a buffer and unmap it you’ll get the promise back with the shmem. And we’d have to then send it back and say “just kidding”.
* JB: We decided not to do this. Would have let us use some prebuilt shmem ownership tracking but decided it would be too baroque. Now using “unsafe shmem”.
* KN: Thought about this a bunch, I’m pretty sure that it’s not really possible to store *no* state on the client. The client needs to know certain things.
* MM: I think it’s possible to be somewhat simpler via tracking generation ids for these exchanges.
* KN: *different* client side state, but still have some.
* JB: Corentin said one of the reasons for behavior being specified the way it is is wanting to allow for post-v1 multiple mapAsync regions on the same buffer. It would be possible to have multiple regions mapped. In this proposal #3014, the state reported would be misleading.
* KN: I think there is room for support for sub-range mapping, but it gets a little strange, e.g. “pending” with some-pending-some-unmapped sub-ranges. But we’d also probably need a new e.g. Unmap() function too, since right now that’s only for the whole buffer.
* MM: Could have multi-map be a new thing that just exposes a “multi-mapped” state.
* JB: Then if they called multi-map they could get that state. If they don’t then they can’t get the new state.
* KN: MM’s suggestion makes sense. Think we can always make it work somehow.
* KN: “mappable buffer views” were another proposed API for this and wouldn’t have as much of an issue.
* …
* KN: Note multi-mapping would be in a world where you can also *use* parts of the buffer that aren’t mapped.
* KG: … Consensus was basically we had better things to do for V1.
* **(Let’s come back to this next week)**

</details>

**Comment by kainino0x** [comment]
Agreed but adding to tacit resolution for @litherum to give input at the next meeting (or earlier)

**Comment by kainino0x** [comment]
Meeting: Resolved to accept this

I'll figure out the landing of this and #3300 todayish.

**Review by kdashg — APPROVED** [review]

**Comment by kainino0x** [comment]
was only waiting for approval from Apple, and got explicit approval so landing now.

**Pull request merged**
